File: round_trip.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

bool check = false;
int n, m;

void dfs(int start, vector<bool> &vis, vector<vector<int>> &adj) {
    if (check) return;
    stack<int> s;
    s.push(start);
    vector<int> path, prev(n + 1, -1);
    while(!s.empty()) {
        if (check) return;
        int curr = s.top();
        s.pop();
        vis[curr] = true;
        prev[curr] = path.size();
        path.push_back(curr);
        
        for (int next : adj[curr]) {
            if (!vis[next]) s.push(next);
            else {
                if (prev[next] == -1) continue;

                if (path.size() - prev[next] >= 3) {
                    check = true;
                    cout << path.size() - prev[next] + 1 << endl;
                    for (int i = prev[next]; i < path.size(); i++) {
                        cout << path[i] << " ";
                    }
                    cout << path[prev[next]] << endl;
                }
            }
        }
    }
}

int main(void) {
    cin >> n >> m;
    vector<vector<int>> adj(n + 1, vector<int>());

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
    vector<bool> vis(n + 1, false);
    for (int i = 1; i <= n; i++) {
        if (!vis[i]) dfs(i, vis, adj);
        if (check) break;
    }
    if (!check) cout << "IMPOSSIBLE" << endl;
}

File: roundTrip.java
import java.io.*;
import java.util.*;

public class roundTrip {
    
    static boolean check = false;
    private static void dfs(int curr, boolean[] vis, List<List<Integer>> adj, List<Integer> path, int[] prev) {
        if (check) return;
        vis[curr] = true;
        prev[curr] = path.size();
        path.add(curr);

        for (int next : adj.get(curr)) {
            if (!vis[next]) {
                dfs(next, vis, adj, path, prev);
            } else {
                int idx = prev[next];
                if (idx == -1) continue;

                // 1 2 3 4 5
                if (path.size() - idx + 1 > 3) {
                    check = true;
                    System.out.println(path.size() - idx + 1);
                    for (int i = idx; i < path.size(); i++) {
                        System.out.print(path.get(i) + " ");
                    }
                    System.out.println(path.get(idx));
                    return;
                }
            }
        }
        path.remove(path.size() - 1);
        prev[curr] = -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i <= n; i++) adj.add(new ArrayList<>());
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            adj.get(a).add(b);
            adj.get(b).add(a);
        }
        boolean vis[] = new boolean[n + 1];
        for (int i = 1; i <= n; i++) {
            int[] prev = new int[n + 1];
            Arrays.fill(prev, -1);
            if (!vis[i]) dfs(i, vis, adj, new ArrayList<>(), prev);
            if (check) break;
        }
        if (!check) System.out.println("IMPOSSIBLE");
    }
}

